plutus-ledger-api/src/PlutusLedgerApi/V3/Contexts.hs

```diff
@@ -0,0 +1,581 @@
+  deriving (Pretty) via (PrettyShow GovernanceAction)
+
+instance PlutusTx.Eq GovernanceAction where
+  {-# INLINEABLE (==) #-}
+  ParameterChange == ParameterChange = Haskell.True
+  HardForkInitiation a == HardForkInitiation a' = a PlutusTx.== a'
+  TreasuryWithdrawals a == TreasuryWithdrawals a' = a PlutusTx.== a'
+  NoConfidence == NoConfidence = Haskell.True
+  NewCommittee a b == NewCommittee a' b' =
+    a PlutusTx.== a' PlutusTx.&& b PlutusTx.== b'
+  NewConstitution a == NewConstitution a' = a PlutusTx.== a'
+  InfoAction == InfoAction = Haskell.True
+  _ == _ = Haskell.False
+
+data ProposalProcedure = ProposalProcedure
+  { ppDeposit          :: V2.Value
+  , ppReturnAddr       :: V2.Credential
+  , ppGovernanceAction :: GovernanceAction
+  , ppAnchor           :: Anchor
+  }
+  deriving stock (Generic, Haskell.Show, Haskell.Eq)
+
+instance Pretty ProposalProcedure where
+  pretty ProposalProcedure{..} =
+    vsep
+      [ "ppDeposit:" <+> pretty ppDeposit
+      , "ppReturnAddr:" <+> pretty ppReturnAddr
+      , "ppGovernanceAction:" <+> pretty ppGovernanceAction
+      , "ppAnchor:" <+> pretty ppAnchor
+      ]
+
+instance PlutusTx.Eq ProposalProcedure where
+  {-# INLINEABLE (==) #-}
+  ProposalProcedure a b c d == ProposalProcedure a' b' c' d' =
+    a PlutusTx.== a'
+      PlutusTx.&& b PlutusTx.== b'
+      PlutusTx.&& c PlutusTx.== c'
+      PlutusTx.&& d PlutusTx.== d'
+
+data ScriptPurpose
+  = Minting V2.CurrencySymbol
+  | Spending V2.TxOutRef
+  | Rewarding V2.Credential
+  | Certifying TxCert
+  | Voting Voter GovernanceActionId
+  | -- TODO: this may turn out to be unused.
+    Proposing
+  deriving stock (Generic, Haskell.Show, Haskell.Eq)
+  deriving (Pretty) via (PrettyShow ScriptPurpose)
+
+instance PlutusTx.Eq ScriptPurpose where
+  {-# INLINEABLE (==) #-}
+  Minting a == Minting a' =
+    a PlutusTx.== a'
+  Spending a == Spending a' =
+    a PlutusTx.== a'
+  Rewarding a == Rewarding a' =
+    a PlutusTx.== a'
+  Certifying a == Certifying a' =
+    a PlutusTx.== a'
+  Voting a b == Voting a' b' =
+    a PlutusTx.== a' PlutusTx.&& b PlutusTx.== b'
+  Proposing == Proposing = Haskell.True
+  _ == _ = Haskell.False
+
+-- | TxInfo for PlutusV3
+data TxInfo = TxInfo
+  { txInfoInputs                :: [V2.TxInInfo]
+  , txInfoReferenceInputs       :: [V2.TxInInfo]
+  , txInfoOutputs               :: [V2.TxOut]
+  , txInfoFee                   :: V2.Value
+  , txInfoMint                  :: V2.Value
+  , txInfoTxCerts               :: [TxCert]
+  , txInfoWdrl                  :: Map V2.Credential Haskell.Integer
+  , txInfoValidRange            :: V2.POSIXTimeRange
+  , txInfoSignatories           :: [V2.PubKeyHash]
+  , txInfoRedeemers             :: Map ScriptPurpose V2.Redeemer
+  , txInfoData                  :: Map V2.DatumHash V2.Datum
+  , txInfoId                    :: V2.TxId
+  , txInfoVotingProcedures      :: Map Voter (Map GovernanceActionId VotingProcedure)
```

Comment:
It's not totally obvious to me that it's best to represent this as a map rather than a flat list with the Voter and GovernanceID included in VotingProcedure?

Reply:
Well on the ledger side it has to be a nested Map for correctness and efficiency.

* Order of votes does not matter, so we don't need it to be a list
* We ensure there are no duplicate votes in the same transaction, for that (Voter, GovernanceActionId) pair must be unique, hence the usage of Map.
* It is a nested Map, because it is a more efficient representation on the wire (eg. a single DRep making many votes in the same transaction, then we only encode Voter once, rather than as many times as there are votes). And it also makes the rules implementation easier as well.

In any case, those are Ledger related reasons for the structure. For Plutus, you'll know better how it should look like. We can just as easily convert it into a flat list of three tuples if you want:

```haskell
 , txInfoVotingProcedures      :: [Voter, GovernanceActionId, VotingProcedure]
```
or a flat Map:
```haskell
 , txInfoVotingProcedures      :: Map (Voter, GovernanceActionId) VotingProcedure
```

Doesn't matter to me.

Reply:
I think it's fine this way, especially if that's how it is in the ledger. It just occured to me that we don't have a very clear rationale for why some of these fields are "indexed" by something and others aren't.